proposals/user-scripts-api.md

```diff
@@ -0,0 +1,189 @@
+
+4. **Introduce a user script world with custom CSP.**  Allow user scripts to opt-in to a more secure world that is only available in this namespace.
+
+### API Schema
+
+#### Types
+
+```
+dictionary RegisteredUserScript {
+  boolean? allFrames;
+  ScriptSource[] js;
+  string[]? excludeMatches;
+  string id;
+  string[]? matches;
+  // Default to `document_idle`
+  RunAt runAt;
+  // Allows `USER_SCRIPT` (default) and `MAIN`
+  // and returns error for `ISOLATED`.
+  ExecutionWorld? world;
+  // Implemented as disjunctions.
+  string[]? includeGlobs;
+  string[]? excludeGlobs;
+}
+dictionary UserScriptFilter {
+  string[]? ids;
+}
+// Errors if both code and file are specified.
+dictionary ScriptSource {
+  string code;
+  string file;
+}
+```
+
+#### Methods
+
+```
+browser.userScripts.register(
+  scripts: RegisteredUserScript[],
+  callback?: function
+)
+browser.userScripts.unregister(
+  filter?: UserScriptFilter[],
+  callback?: function
+)
+browser.userScripts.getScripts(
+  filter?: UserScriptFilter[],
+  callback?: function
+)
+browser.userScripts.update(
+  scripts?: RegisteredUserScript[],
+  callback?: function
+)
+```
+
+### Requirements
+
+#### A. A mechanism to execute code in the main world
+
+- User scripts can be registered in the `MAIN` world.
+- Extension can customize the `USER_SCRIPT` world's CSP to inject a script tag into the host page (expanded in requirement B)
+
+#### B. The ability to execute code (with a separate CSP) in a world different from the main world and the extension's isolated world
+
+##### `USER_SCRIPT` World
+
+- Exempt from the page's CSP, can customize its own CSP (e.g can allow injection of `unsafe-inline`).
+- Can communicate with the extension using extension messaging APIs (expanded under [Messaging](#Messaging)).
+- Isolated from the web page (similar to other isolated worlds), but will be largely un-permissioned. It will not have access to any extension APIs, except messaging APIs mentioned, and will not have any cross-origin exceptions (these don't exist in Chrome, but do in other browsers).  In the future, we may bring more APIs (such as chrome.dom APIs or a dedicated API to execute a script in the main world), but these APIs will not grant any additional privilege to the script beyond what it has to access the page content.
+- Can communicate with different JS worlds via `window.postMessage()`. A dedicated API to communicate between worlds is being considered as part of future work.
+- Shares DOM with the web page.  Code from both worlds cannot directly interact with each other, except through DOM APIs.
+- When an asymmetric security relationship may exist, the `MAIN` world is considered to be less privileged than the `USER_SCRIPT` world
+
+##### Configuration
+
+```
+browser.userScripts.configureWorld(
+  csp?: string
+  enableMessaging?: bool
+)
```

Comment:
Also, it should be just `messaging` because it's not a verb/method but just an option, same as `csp` is not `setCSP`.

Reply:
Option names are not prefixed with "enable" unless it's an action. In the extensions API there are only a couple of such cases in DNR (enableRulesetIds, enableRuleIds) because there is a complementary `disable` action'y option.

And I see Chromium devs have already made the same mistake in [enableGranularPermissions](https://developer.chrome.com/docs/extensions/reference/identity/#type-TokenDetails).

Non-actionable options may have `Enabled` suffix, although it's also an unnecessary fluff.